### PR TITLE
Break assignment whose rhs is wrapped by an unary expression

### LINKED ISSUE #9490: Unary expressions prevent line break
Although printWidth is set to `80` but some times Prettier don't break lines to shorter lines so my linter complains that `This line has a length of X. Maximum allowed is 81. eslint(max-len)` I have seen this issue in lines with long name variables. 

**Prettier 2.1.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAzArlMMCW0AEAEnADYkQDqEATiQCYAUwqNAtgM4AKJAhmHKwQwAynBi4oAcwA0AXwCU+YAB0o+fJCjsY+aCQCeAJQR041YwEd0OanDrD0AI1Y527PFvwBefAEIW1BzcfAJCouI4UgB0AIKcnADyAJIAcgAqALIAoukA+oZZAGpJWRQA3KrqAPRV+Bk0cPgkkXDsuqgaEKZRPaqyINIgEAAOuNDsyKA81NQQAO6c0wgTKDwkczz6E4OO1HwA1mLCw3yRksgw1OhwgwKOdqZ0ADI8Uug8knAAYmw8EVLIEA8dAwCADEAACxgrBIFAhOHg7BO-FEWgROAAbgj9ICwG5wZF2GYYJw9pJWDxkKg1kTBgArdgADwAQnswIcRDxBE8WlSaTcQAzGcIziQ4ABFdAQeB8ki0kAnahE6iAxw8e4kcHDaiRGAUHB0GAQ5AADgADINtRAiRQ9sNAdrWmYMXBwVZpXBSSMVkD2ABaKBwOx2cG2Kw2T1kimy+VElwXK4C9xSMWSj0xgUwdX6w3GpAAJkGlx4OGaUgAwhBWNGUK0AKzg9BEtLqlbUuUCjHXJJQUywYRgHWjGK94QwfRijOyWRAA)
```sh
--parser babel
```

**options:**
```json
{
  "arrowParens": "always",
  "bracketSpacing": true,
  "embeddedLanguageFormatting": "auto",
  "htmlWhitespaceSensitivity": "css",
  "insertPragma": false,
  "jsxBracketSameLine": false,
  "jsxSingleQuote": false,
  "printWidth": 80,
  "proseWrap": "preserve",
  "quoteProps": "as-needed",
  "requirePragma": false,
  "semi": true,
  "singleQuote": false,
  "tabWidth": 2,
  "trailingComma": "es5",
  "useTabs": false,
  "vueIndentScriptAndStyle": false
}
```

**Input:**
```jsx
function HelloWorld({formsPlacementSetting,}) {
  const onlyRenderRequiredSubmissions = !formsPlacementSetting.APPOINTMENT_REVIEW;
  // More lines of code...
}
```

**Output:**
```jsx
function HelloWorld({ formsPlacementSetting }) {
  const onlyRenderRequiredSubmissions = !formsPlacementSetting.APPOINTMENT_REVIEW;
  // More lines of code...
}

```

**Expected behavior:**
```jsx
function HelloWorld({ formsPlacementSetting }) {
  const onlyRenderRequiredSubmissions = 
    !formsPlacementSetting.APPOINTMENT_REVIEW;
  // More lines of code...
}

```

### PR DESCRIPTION
**Pull request by sosukesuzuki**
## Description

Fixes #9490

## Checklist

- [x] I’ve added tests to confirm my change works.
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by fisker** [comment]
**Prettier pr-10158**
[Playground link](https://deploy-preview-10158--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuc0DOMAEAbCf8GFGFQDmAzJgLwA6UmmAhLscWQHQusmmdFkBuOpCgYc3CWwrU6DRs0kEOXRcv6khUEWJWKJZSrXqYAQnmxwAhlAAUu-b3vc1PAJQCQAGhAQADjABLdGRQSwAnMIgAdwAFcIQ0ZBBLbCjLAE9E7wAjMMswAGs4GABlX3yAsmQYMIBXOG84AFtsuAATNvaAGWtSWstSOAAxCDCmyxhAqpRLWpgILxAACxgm7AB1JYD4NHKwOBKE7YCAN230pLA0LJBKtDgwmBi80nHkADMU++8AKzQADxMeUKxRKliacC6lTgHy+DRAf3+JUqpAsAEVahB4LDsN8QOUwvcwklspZWthFr4wpUYOsAm0YEtkAAOAAM3ipEHu6zyviSVLgRJOMO8AEdMfBnn5EjM0ABaKBwdrtRZhODigJq54DN5IT64+H3JoBap1Q0o9ESmF6uHeGBkukMplIABMdryAWwKIAwhAmrqQIKAKyLWr3AAqZJl+rxJ3qAEkoJ1YCUwNT-ABBJMlGDpCw4+4AX0LQA)
```sh
--parser babel
```

**Input:**
```jsx
const loooooooooooooooooooooooooong3 =
  !looooooooooooooong.looooooooooooooong.loooooong;
const loooooooooooooooooooooooooong3 =
  !!looooooooooooooong.looooooooooooooong.loooooong;
const loooooooooooooooooooooooooong3 =
  Boolean(looooooooooooooong.looooooooooooooong.loooooong);
```

**Output:**
```jsx
const loooooooooooooooooooooooooong3 =
  !looooooooooooooong.looooooooooooooong.loooooong;
const loooooooooooooooooooooooooong3 = !!looooooooooooooong.looooooooooooooong
  .loooooong;
const loooooooooooooooooooooooooong3 = Boolean(
  looooooooooooooong.looooooooooooooong.loooooong
);

```

File: src/language-js/print/assignment.js [hunk 1]
```diff
@@ -63,6 +67,17 @@ function printAssignmentRight(leftNode, rightNode, printedRight, options) {
...
+   * @returns {Node}
+   */
+  function getUnaryArgument(node) {
+    if (node.type === "UnaryExpression") {
+      return node.argument;
+    }
+    return node;
+  }
+
   const canBreak =
```
Comment (fisker) [inline src/language-js/print/assignment.js:81]: Let's move this out to a function in this file.

File: src/language-js/print/assignment.js [hunk 2]
```diff
@@ -67,18 +67,26 @@ function printAssignmentRight(leftNode, rightNode, printedRight, options) {
...
+ * @returns {Node}
+ */
+function getUnaryArgument(node) {
+  if (node.type === "UnaryExpression") {
+    return getUnaryArgument(node.argument);
+  }
+  return node;
+}
+
+function canBreakAssignmentRight(leftNode, rightNode, options) {
```
Comment (fisker) [inline src/language-js/print/assignment.js:88]: I mean rewrite in `if-else`, hard to read.WDYT?

File: src/language-js/print/assignment.js [hunk 3]
```diff
@@ -67,18 +67,26 @@ function printAssignmentRight(leftNode, rightNode, printedRight, options) {
...
 
-  const canBreak =
+  return [" ", printedRight];
+}
+
+/**
+ * @param {Node} node
+ * @returns {Node}
+ */
+function getUnaryArgument(node) {
```
Comment (fisker) [inline src/language-js/print/assignment.js:81]: We can use a loop here, I'm not sure if we should only check `!`, `!+a` `+!a` seems very uncommon.

File: src/language-js/print/assignment.js (hunk 2 above)
Comment (sosukesuzuki) [reply src/language-js/print/assignment.js:88]: like this?

```js
function canBreakAssignmentRight(leftNode, rightNode, options) {
  if (isBinaryish(rightNode) && !shouldInlineLogicalExpression(rightNode)) {
    return true;
  }
  if (
    rightNode.type === "ConditionalExpression" &&
    isBinaryish(rightNode.test) &&
    !shouldInlineLogicalExpression(rightNode.test)
  ) {
    return true;
  }
  if (rightNode.type === "StringLiteralTypeAnnotation") {
    return true;
  }
  if (
    rightNode.type === "ClassExpression" &&
    isNonEmptyArray(rightNode.decorators)
  ) {
    return true;
  }
  if (
    (leftNode.type === "Identifier" ||
      isStringLiteral(leftNode) ||
      leftNode.type === "MemberExpression") &&
    (isStringLiteral(getUnaryArgument(rightNode)) ||
      isMemberExpressionChain(getUnaryArgument(rightNode))) &&
    // do not put values on a separate line from the key in json
    options.parser !== "json" &&
    options.parser !== "json5"
  ) {
    return true;
  }
  if (rightNode.type === "SequenceExpression") {
    return true;
  }
  return false;
}
```

File: src/language-js/print/assignment.js (hunk 3 above)
Comment (sosukesuzuki) [reply src/language-js/print/assignment.js:81]: > We can use a loop here.

Sounds good.

>  `!+a` `+!a` seems very uncommon.

I think so too. But I don't think it's necessary to add a condition to exclude it.

File: src/language-js/print/assignment.js (hunk 2 above)
Comment (fisker) [reply src/language-js/print/assignment.js:88]: More blank lines.

Comment (fisker) [reply src/language-js/print/assignment.js:88]: And less ternary, still a lot.

Comment (sosukesuzuki) [reply src/language-js/print/assignment.js:88]: Hmm, I have no idea to less logical operators cleanly...

**Comment by thorn0** [comment]
Why did this change? see https://github.com/thorn0/prettier-regression-testing/issues/2#issuecomment-768260063
```diff
 function isSourceErrorModule(currentFilePath, targetModulePath, src) {
-  for (const srcPath of [src, `${src}.js`, `${src}/index`, `${src}/index.js`]) {
+  for (const srcPath of [
+    src,
+    `${src}.js`,
+    `${src}/index`,
+    `${src}/index.js`,
+  ]) {
```

**Comment by thorn0** [comment]
Disregard. I broke that in #10106.

**Comment by brodybits** [comment]
I tried a few cases related to #2482 which seem to be still not fixed:

**Prettier pr-10158**
[Playground link](https://deploy-preview-10158--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuc0DOMAEAbCf8GFGFQDmAzJgLwA6UmmAhLscWS6yaR-mQNx06kKBhydxbCtToNmEgu3m9uRMgAoAlAKhD0WHks5lKtepgDkB8YsM2FpcwDpsCUjAAWgqMNFXDqqVNZP1YyZ38wq3UeMi06EAAaEAgABxgAS3RkUABDACc8iAB3AAV8hDRkEBzsIpyAT0qkgCM8nLAAazgYAGUU9vSyZBg8gFc4JLgAW2a4ABM5+YAZHLJRnNI4ADEIPKmcmAyhlBzRmAhEkHcYKewAdXd0+DR+sDgeiqf0gDcn+qqwGgmiBBmg4HkYCU2qR9sgAGY1MFJABWaAAHgAhNqdbo9HJTOBLQZweGIiYgVFonqDbhwACKowg8FJ2CRIH6eTBeSqzRys2wlxSeUGMDu6TmHmQAA4AAxJIUQMF3NopKpCuBc74kpIAR0Z8ChqUqJzQAFooHB5vNLnk4Hr0raoRtYUgEazyWCpulhmMPTSXAymSTXWSkjA+WKJe5kAAmMNtdLYGkAYQgUxdIA1AFZLqMwQAVPnGt1s77jACSUEWsB6YGFaQAglWejB6i4WWCAL6doA)
```sh
--parser babel
```

**Input:**
```jsx
const loooooooooooooooooooooooooong3 =
  !looooooooooooooonglooooooooooooooongloooooong;

const loooooooooooooooooooooooooong3 =
  !looooooooooooooonglooooooooooooooongloooooong();

const loooooooooooooooooooooooooong3 =
  'looooooooooooooonglooooooooooooooongloooooong'.length

const loooooooooooooooooooooooooong3 =
  !looooooooooooooong.looooooooooooooong.loooooong(loooong);

```

**Output:**
```jsx
const loooooooooooooooooooooooooong3 = !looooooooooooooonglooooooooooooooongloooooong;

const loooooooooooooooooooooooooong3 = !looooooooooooooonglooooooooooooooongloooooong();

const loooooooooooooooooooooooooong3 = "looooooooooooooonglooooooooooooooongloooooong"
  .length;

const loooooooooooooooooooooooooong3 = !looooooooooooooong.looooooooooooooong.loooooong(
  loooong
);

```

**Comment by thorn0** [comment]
@brodybits The scope of this PR is consistency of the output with and without `!`/`-`.

**Comment by sosukesuzuki** [comment]
> I tried a few cases related to #2482 which seem to be still not fixed:

Yes, but these are not related unary.

**Prettier 2.2.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEAbCf8GFGFQDmAzJgLwA6UmOxTZuTRL7pA3HXZFBkZthnSrXpCReDlOmlWBMgAoAlDyh90WBbOFkxdBgHIdImbPP4yRgHTYEpGAAteUfoNO7i+6oclSyOy9Az2UdMjU6EAAaEAgABxgAS3RkUABDACdMiAB3AAUshDRkEHTsXPSATxLYgCNM9LAAazgYAGV4pqSyZBhMgFc4WLgAWzq4ABNJqYAZdLIB9NI4ADEITNH0mGTelHSBmAgYkCcYUewAdSck+DQusDh24tukgDdbqtKwNFqQHrQcEyMHyjVIW2QADNyoDYgArNAADwAQo0Wm12ulRnBZj04FCYcMQAjEe0evI4ABFAYQeAE7CwkBdTKAzKlOrpCbYE7xTI9GCXJKTZzIAAcAAZYryIIDLo14qVeXBWW98bEAI40+CghIlfZoAC0UDgUymJ0ycE1SQtoOWEKQ0IZRMBoySfUGzvJ9mptPxDsJsRgnMFwqcyAATIHGklsOSAMIQUb2kDKgCsJwGgIAKpy9Y7GW8hgBJKAzWDtMB8xIAQVL7RgVXs9MBAF8W0A)
```sh
--parser babel
```

**Input:**
```jsx
const loooooooooooooooooooooooooong3 =
  looooooooooooooonglooooooooooooooongloooooong;

const loooooooooooooooooooooooooong3 =
  looooooooooooooonglooooooooooooooongloooooong();

const loooooooooooooooooooooooooong3 =
  'looooooooooooooonglooooooooooooooongloooooong'.length

const loooooooooooooooooooooooooong3 =
  looooooooooooooong.looooooooooooooong.loooooong(loooong);

```

**Output:**
```jsx
const loooooooooooooooooooooooooong3 = looooooooooooooonglooooooooooooooongloooooong;

const loooooooooooooooooooooooooong3 = looooooooooooooonglooooooooooooooongloooooong();

const loooooooooooooooooooooooooong3 = "looooooooooooooonglooooooooooooooongloooooong"
  .length;

const loooooooooooooooooooooooooong3 = looooooooooooooong.looooooooooooooong.loooooong(
  loooong
);

```

File: src/language-js/print/assignment.js (hunk 3 above)
Comment (fisker) [reply src/language-js/print/assignment.js:81]: If you agree it's not common, I think we should add one more condition, simple

```js
const node =
  node.type === "UnaryExpression" &&
  node.operator === "!" &&
  node.argument.type === "UnaryExpression" &&
  node.argument.operator === "!"
    ? node.argument.argument
    : node.argument;
```

Only any `UnaryOperator` + `!!`

Comment (fisker) [reply src/language-js/print/assignment.js]: This function confused me... I thought it's always `UnaryExpression`, I wounder why I keep failing on write these checks... It also allow `StringLiteral`

Comment (fisker) [reply src/language-js/print/assignment.js]: I'm done with rewriting `canBreakAssignmentRight`, I think it is more clear now.

I removed `getUnaryArgument` function, as it's just one line, and this function name is confusing. 

I'll let you decide if accept my suggestion.

**Review by thorn0 — APPROVED** [review]

**Comment by sosukesuzuki** [comment]
@fisker Thank you.

**Pull request merged**